Mathlib/LinearAlgebra/FiniteDimensional.lean

```diff
@@ -1365,51 +1410,66 @@ alias Subalgebra.finiteDimensional_toSubmodule ↔
 #align finite_dimensional.of_subalgebra_to_submodule FiniteDimensional.of_subalgebra_toSubmodule
 #align finite_dimensional.subalgebra_to_submodule FiniteDimensional.subalgebra_toSubmodule
 
+set_option synthInstance.etaExperiment true in
 instance FiniteDimensional.finiteDimensional_subalgebra [FiniteDimensional F E]
     (S : Subalgebra F E) : FiniteDimensional F S :=
   FiniteDimensional.of_subalgebra_toSubmodule inferInstance
 #align finite_dimensional.finite_dimensional_subalgebra FiniteDimensional.finiteDimensional_subalgebra
 
+set_option synthInstance.etaExperiment true in
 instance Subalgebra.finiteDimensional_bot : FiniteDimensional F (⊥ : Subalgebra F E) := by
   nontriviality E
   exact finiteDimensional_of_rank_eq_one Subalgebra.rank_bot
 #align subalgebra.finite_dimensional_bot Subalgebra.finiteDimensional_bot
 
-theorem Subalgebra.eq_bot_of_rank_le_one {S : Subalgebra F E} (h : Module.rank F S ≤ 1) : S = ⊥ :=
-  by
+set_option synthInstance.etaExperiment true in
+set_option maxHeartbeats 300000 in
+theorem Subalgebra.eq_bot_of_rank_le_one {S : Subalgebra F E} (h : Module.rank F S ≤ 1) :
+    S = ⊥ := by
   nontriviality E
-  obtain ⟨m, hm, he⟩ := Cardinal.exists_nat_eq_of_le_nat (h.trans_eq nat.cast_one.symm)
-  haveI := finiteDimensional_of_rank_eq_nat he
-  rw [← not_bot_lt_iff, ← subalgebra.to_submodule.lt_iff_lt]
-  haveI := S.to_submodule_equiv.symm.FiniteDimensional
-  refine' fun hl => (Submodule.finrank_lt_finrank_of_lt hl).not_le (nat_cast_le.1 _)
+  obtain ⟨m, _, he⟩ := Cardinal.exists_nat_eq_of_le_nat (h.trans_eq Nat.cast_one.symm)
+  -- porting note: fails without explicit type
+  haveI : FiniteDimensional F S := finiteDimensional_of_rank_eq_nat he
+  rw [← not_bot_lt_iff, ← Subalgebra.toSubmodule.lt_iff_lt]
+  -- porting note: fails without explicit type
+  haveI : FiniteDimensional F (Subalgebra.toSubmodule S) :=
+    S.toSubmoduleEquiv.symm.finiteDimensional
+  refine fun hl => (Submodule.finrank_lt_finrank_of_lt hl).not_le (natCast_le.1 ?_)
   iterate 2 rw [Subalgebra.finrank_toSubmodule, finrank_eq_rank]
-  exact h.trans_eq subalgebra.rank_bot.symm
+  exact h.trans_eq Subalgebra.rank_bot.symm
 #align subalgebra.eq_bot_of_rank_le_one Subalgebra.eq_bot_of_rank_le_one
 
 theorem Subalgebra.eq_bot_of_finrank_one {S : Subalgebra F E} (h : finrank F S = 1) : S = ⊥ :=
   Subalgebra.eq_bot_of_rank_le_one <| by
-    haveI := finite_dimensional_of_finrank_eq_succ h
+    -- porting note: fails without explicit type
+    haveI : FiniteDimensional F S := finiteDimensional_of_finrank_eq_succ h
     rw [← finrank_eq_rank, h, Nat.cast_one]
+    -- porting note: added, `rw` forgot to close this
+    exact le_rfl
```

Comment:
I assume `rw` is doing "`rfl`, but with a short timeout". This seems like a questionable design, because changes to performance can then change the goal state.